File: ios/Sources/Prefs.swift
import Foundation

/// Persisted settings, mirroring the Android `Prefs` (UserDefaults instead of
/// SharedPreferences). The device id is derived once and kept forever.
enum Prefs {
    private static let d = UserDefaults.standard
    private enum Key {
        static let host = "host"
        static let controlHost = "control_host"
        static let port = "port"
        static let enabled = "enabled"
        static let deviceID = "device_id"
        static let ingestToken = "ingest_token"
    }

    /// Isis (the fleet control plane) over WireGuard: a stable address, so it's the
    /// out-of-the-box default and existing installs self-heal without reconfiguration.
    /// The stream still goes to the recorder `host`; only the API moved here.
    static let defaultControlHost = "10.100.0.2"

    /// Recorder host the PCM stream connects to (the Mac's ingest, on the home LAN).
    /// Empty until the user sets it.
    static var host: String {
        get { d.string(forKey: Key.host) ?? "" }
        set { d.set(newValue, forKey: Key.host) }
    }

    /// Control-plane host for the capture API — pause/resume and the fleet liveness the
    /// Devices panel shows. That's Isis, not the recorder host: the Isis split put the API
    /// and the PCM ingest on different machines. Empty/unset falls back to
    /// `defaultControlHost`, so the controls and panel work out of the box.
    static var controlHost: String {
        get {
            let h = d.string(forKey: Key.controlHost) ?? ""
            return h.isEmpty ? defaultControlHost : h
        }
        set { d.set(newValue, forKey: Key.controlHost) }
    }

    /// recalld's ingest plane (docs/architecture.md): same host as the control
    /// plane, its own port. Not user-set until a reason appears.
    static let ingestBase = "http://10.100.0.2:8001"

    /// The fourth credential plane's per-device bearer: `PUT` this phone's own
    /// segments to recalld, and nothing else. Empty = send no header.
    static var ingestToken: String {
        get { d.string(forKey: Key.ingestToken) ?? "" }
        set { d.set(newValue, forKey: Key.ingestToken) }
    }

    /// Shared ingest port (matches the recorder's DEFAULT_INGEST_PORT).
    static var port: Int {
        get {
            let p = d.integer(forKey: Key.port)
            return p == 0 ? 9999 : p
        }
        set { d.set(newValue, forKey: Key.port) }
    }

    /// Whether the user last pressed Start — drives auto-resume on launch.
    static var enabled: Bool {
        get { d.bool(forKey: Key.enabled) }
        set { d.set(newValue, forKey: Key.enabled) }
    }

    /// Pre-set fixed source id for this dedicated device — like the Pixels' `pixel9`/
    /// `pixel5`. Set to nil to auto-derive `<model>-<random>` on first run instead.
    static let presetID: String? = "iphone11"

    /// Stable source id; the pre-set value if any, else generated and stored on first read.
    static var deviceID: String {
        if let preset = presetID {
            if d.string(forKey: Key.deviceID) != preset { d.set(preset, forKey: Key.deviceID) }
            return preset
        }
        if let existing = d.string(forKey: Key.deviceID), !existing.isEmpty {
            return existing
        }
        let fresh = DeviceID.generate()
        d.set(fresh, forKey: Key.deviceID)
        return fresh
    }
}

File: android/app/src/main/kotlin/org/recall/mic/Prefs.kt
package org.recall.mic

import android.content.Context
import android.os.Build
import java.util.UUID

/**
 * Persisted config for the mic app, read by the foreground service and the boot
 * receiver so they share one configuration. Two hosts, because the Isis split put the
 * PCM ingest and the control API on different machines:
 *  - [host] — the *recorder* host the mic stream connects to (the Mac's ingest, on the
 *    home LAN); user-set, per-network.
 *  - [controlHost] — the *control-plane* host for the capture API (pause/resume + the
 *    fleet liveness the Devices panel shows). That's Isis, reachable over the VPN, so it
 *    defaults to [DEFAULT_CONTROL_HOST] and rarely changes.
 * The ingest port is fixed (one shared port) and the device id is derived.
 */
object Prefs {
    private const val FILE = "recall-mic"
    private const val KEY_HOST = "host"
    private const val KEY_CONTROL_HOST = "control_host"
    private const val KEY_DEVICE_ID = "device_id"

    // Legacy key from the manual-device-id version; adopted for source continuity.
    private const val KEY_LEGACY_SOURCE_ID = "source_id"
    private const val KEY_ENABLED = "enabled"
    private const val MAX_ID_LEN = 40

    // Isis (the fleet control plane) over WireGuard: a stable address, so it's the
    // out-of-the-box default and existing installs self-heal without reconfiguration.
    // The stream still goes to the recorder [host]; only the API moved here.
    const val DEFAULT_CONTROL_HOST = "10.100.0.2"

    private fun prefs(ctx: Context) = ctx.getSharedPreferences(FILE, Context.MODE_PRIVATE)

    private fun sanitize(raw: String): String =
        raw.lowercase().replace(Regex("[^a-z0-9]+"), "-").trim('-')

    fun host(ctx: Context): String = prefs(ctx).getString(KEY_HOST, "") ?: ""

    /** The capture-API host (Isis). Empty/unset falls back to [DEFAULT_CONTROL_HOST], so
     * the pause controls and Devices panel work out of the box against the fleet. */
    fun controlHost(ctx: Context): String =
        (prefs(ctx).getString(KEY_CONTROL_HOST, "") ?: "").ifEmpty { DEFAULT_CONTROL_HOST }

    fun saveControlHost(ctx: Context, controlHost: String) {
        prefs(ctx).edit().putString(KEY_CONTROL_HOST, controlHost).apply()
    }

    /** This device's stable recall source id, announced in the stream handshake.
     * Resolved once and persisted — nothing for the user to set; renamable in the web
     * UI. A phone upgraded from the manual-device-id version keeps that id (so its
     * recording history stays one source); a fresh phone derives model + random suffix
     * (so two same-model phones differ). */
    fun deviceId(ctx: Context): String {
        val existing = prefs(ctx).getString(KEY_DEVICE_ID, null)
        if (existing != null) return existing
        val legacy = prefs(ctx).getString(KEY_LEGACY_SOURCE_ID, null)?.let(::sanitize)
        val id =
            (
                legacy?.takeIf { it.isNotEmpty() }
                    ?: "${sanitize(Build.MODEL).ifEmpty { "device" }}-" +
                    UUID.randomUUID().toString().take(8)
            ).take(MAX_ID_LEN)
        prefs(ctx).edit().putString(KEY_DEVICE_ID, id).apply()
        return id
    }

    /** Whether streaming should be running — set true on Start, drives boot restart. */
    fun enabled(ctx: Context): Boolean = prefs(ctx).getBoolean(KEY_ENABLED, false)

    fun save(ctx: Context, host: String, enabled: Boolean) {
        prefs(ctx)
            .edit()
            .putString(KEY_HOST, host)
            .putBoolean(KEY_ENABLED, enabled)
            .apply()
    }
}
